### PR TITLE
new: Added jina embedding v3

### PR DESCRIPTION
**Pull request by hh-space-invader**
To compute canonical vectors:

```
import onnxruntime
import numpy as np
from transformers import AutoTokenizer, PretrainedConfig


def onnx_embed(session: onnxruntime.InferenceSession, task_id: int):
	inputs = {
		'input_ids': input_text['input_ids'],
		'attention_mask': input_text['attention_mask'],
		'task_id': np.array(task_id, dtype=np.int64)
	}
	outputs = session.run(None, inputs)[0]
	embeddings = mean_pooling(outputs, input_text["attention_mask"])
	embeddings = embeddings / np.linalg.norm(embeddings, ord=2, axis=1, keepdims=True)
	return embeddings

def mean_pooling(model_output: np.ndarray, attention_mask: np.ndarray):
    token_embeddings = model_output
    input_mask_expanded = np.expand_dims(attention_mask, axis=-1)
    input_mask_expanded = np.broadcast_to(input_mask_expanded, token_embeddings.shape)
    sum_embeddings = np.sum(token_embeddings * input_mask_expanded, axis=1)
    sum_mask = np.clip(np.sum(input_mask_expanded, axis=1), a_min=1e-9, a_max=None)
    return sum_embeddings / sum_mask

docs = [
    "Hello World",
    "Follow the white rabbit."
]
model_name = 'jinaai/jina-embeddings-v3'
tokenizer = AutoTokenizer.from_pretrained(model_name)

config = PretrainedConfig.from_pretrained(model_name)

input_text = tokenizer(docs, return_tensors='np', padding=True, truncation=True)
input_ids = input_text['input_ids']
attention_mask = input_text['attention_mask']

model_path = 'models/models--jinaai--jina-embeddings-v3/snapshots/62a81741b58448ed8f691764cec7aa5d3c045e4c/onnx/model.onnx'
session = onnxruntime.InferenceSession(model_path)

for task_id in range(5):
	embeddings = onnx_embed(session, task_id)
	print([[round(value, 4) for value in e.tolist()[:5]] for e in embeddings])
```
### All Submissions:

* [ ] Have you followed the guidelines in our Contributing document?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

* [ ] Does your submission pass the existing tests?
* [ ] Have you added tests for your feature?
* [ ] Have you installed `pre-commit` with `pip3 install pre-commit` and set up hooks with `pre-commit install`?

### New models submission:

* [ ] Have you added an explanation of why it's important to include this model?
* [ ] Have you added tests for the new model? Were canonical values for tests computed via the original model?
* [ ] Have you added the code snippet for how canonical values were computed?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: fastembed/text/multitask_embedding.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import numpy as np
+
+from fastembed.text.pooled_normalized_embedding import PooledNormalizedEmbedding
+from fastembed.text.onnx_embedding import OnnxTextEmbeddingWorker
+from fastembed.text.onnx_text_model import TextEmbeddingWorker
+
+supported_multitask_models = [
+    {
+        "model": "jinaai/jina-embeddings-v3",
+        "dim": [32, 64, 128, 256, 512, 768, 1024],
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:12]: We are not introducing matryoshka in this pr, let's make it an int?

**Review by joein — CHANGES_REQUESTED** [review]
1) tests
2) license
3) users can't change task id in this setting, since we are not propagating kwargs to onnx_embed, they are always hard-coded

File: fastembed/text/multitask_embedding.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        "model_file": "onnx/model.onnx",
+        "additional_files": ["onnx/model.onnx_data"],
+    },
+]
+
+
+class JinaEmbeddingV3(PooledNormalizedEmbedding):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._current_task_id = 4
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:35]: let's use a enum like this one and introduce class variable so it would be more flexible

```python
from enum import Enum


class Task(str, Enum):
    RETRIEVAL_QUERY = 0
    RETRIEVAL_PASSAGE = 1
    SEPARATION = 2
    CLASSIFICATION = 3
    TEXT_MATCHING = 4


class JinaEmbeddingV3(PooledNormalizedEmbedding):
    DEFAULT_TASK = Task.TEXT_MATCHING
    PASSAGE_TASK = Task.RETRIEVAL_PASSAGE
    QUERY_TASK = Task.RETRIEVAL_QUERY
```

**Review by joein — CHANGES_REQUESTED** [review]

File: fastembed/text/multitask_embedding.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        "sources": {
+            "hf": "jinaai/jina-embeddings-v3",
+        },
+        "model_file": "onnx/model.onnx",
+        "additional_files": ["onnx/model.onnx_data"],
+    },
+]
+
+
+class Task(str, Enum):
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:33]:
```suggestion
class Task(int, Enum):
```

File: fastembed/text/multitask_embedding.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    ) -> dict[str, np.ndarray]:
+        onnx_input["task_id"] = np.array(self._current_task_id, dtype=np.int64)
+        return onnx_input
+
+    def embed(
+        self,
+        documents: Union[str, Iterable[str]],
+        batch_size: int = 256,
+        parallel: Optional[int] = None,
+        task_id: int = 4,
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:69]:
```suggestion
        task_id: int = DEFAULT_TASK,
```

File: tests/test_text_multitask_embeddings.py [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        embeddings_2 = np.stack(embeddings_2, axis=0)
+
+        embeddings_3 = list(model.embed(docs, batch_size=10, parallel=0))
+        embeddings_3 = np.stack(embeddings_3, axis=0)
+
+        assert embeddings.shape[0] == len(docs) and embeddings.shape[-1] == dim
+        assert np.allclose(embeddings, embeddings_2, atol=1e-4)
+        assert np.allclose(embeddings, embeddings_3, atol=1e-4)
+
+        delete_model_cache(model.model._model_dir)
```
Comment (joein) [inline tests/test_text_multitask_embeddings.py:215]: I think we don't need to test parallel=2 / None / 0, since we are not overriding anything in this class
However, what we need is to check whether we assigning task id properly, e.g. if a user provides task_id=Task.CLASSIFICATION

File: tests/test_text_multitask_embeddings.py [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+            assert embeddings.shape == (len(docs), dim)
+
+            canonical_vector = task["vectors"]
+            assert np.allclose(
+                embeddings[: len(docs), : canonical_vector.shape[1]], canonical_vector, atol=1e-4
+            ), model_desc["model"]
+
+        if is_ci:
+            delete_model_cache(model.model._model_dir)
```
Comment (joein) [inline tests/test_text_multitask_embeddings.py:124]: we also need to test if it works properly with our enum, not only with int

**Review by joein — CHANGES_REQUESTED** [review]

File: fastembed/text/multitask_embedding.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        "model": "jinaai/jina-embeddings-v3",
+        "dim": 1024,
+        "tasks": {
+            "retrieval.query": 0,
+            "retrieval.passage": 1,
+            "separation": 2,
+            "classification": 3,
+            "text-matching": 4,
+        },
+        "description": "Multi-task, multi-lingual embedding model with Matryoshka architecture",
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:21]: Descriptions should contain information about modality, input tokens truncations, prefixes, year of the model, etc

An example (from intfloat/multilingual-e5-large):

`Text embeddings, Unimodal (text), Multilingual (~100 languages), 512 input tokens truncation, Prefixes for queries/documents: necessary, 2024 year.`

**Review by joein — CHANGES_REQUESTED** [review]
lacks tests of propagation task ids in parallel processing

File: tests/test_text_multitask_embeddings.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+    embeddings_1 = np.stack(embeddings_1, axis=0)
+
+    embeddings_2 = list(model.embed(docs, batch_size=10, parallel=1))
+    embeddings_2 = np.stack(embeddings_2, axis=0)
+
+    assert embeddings_1.shape[0] == len(docs) and embeddings_1.shape[-1] == dim
+    assert np.allclose(embeddings_1, embeddings_2, atol=1e-4)
+
+    if is_ci:
+        delete_model_cache(model.model._model_dir)
```
Comment (joein) [inline tests/test_text_multitask_embeddings.py:212]: still lacks a check for propagation of task ids

File: fastembed/text/multitask_embedding.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+        self._current_task_id = self.QUERY_TASK
+
+        if isinstance(query, str):
+            query = [query]
+
+        if not hasattr(self, "model") or self.model is None:
+            self.load_onnx_model()
+
+        for text in query:
+            yield from self._post_process_onnx_output(self.onnx_embed([text]))
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:86]: can we just reuse embed method and pass task_id=self.QUERY_TASK to it?

File: fastembed/text/multitask_embedding.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+class Task(int, Enum):
+    RETRIEVAL_QUERY = 0
+    RETRIEVAL_PASSAGE = 1
+    SEPARATION = 2
+    CLASSIFICATION = 3
+    TEXT_MATCHING = 4
+
+
+class JinaEmbeddingV3(PooledNormalizedEmbedding):
+    DEFAULT_TASK = Task.TEXT_MATCHING
```
Comment (joein) [inline fastembed/text/multitask_embedding.py:42]: I think that it might be inconvenient to have a default task different from passage embed task, could we please make it the same?

**Review by joein — COMMENTED** [review]
Could you please upload scripts for canonical vectors computation to the dedicated repo?

**Review by joein — APPROVED** [review]
